Improve flexibility of Atom feed generator

iay opened the pull request:
Allow overriding `<id>` and `<link rel="alternate">` elements.

### Detailed description

Adds two new parameters, `:alt_link` and `:id`, to the `atom_feed` method. These allow the overriding of the `<id>` and `<link rel="alternate">` elements of the generated Atom feed.

Overriding these can result in greater compatibility and also allow for the possibility of multiple feeds per site.

I have manually tested that if you don't supply these new parameters, you get the same output as before. As mentioned in the comments on #1300, I think there's a separate issue that the default for `:alt_link` should probably be something like `"/"` instead of the base URL, but that seems like a separate issue to this enhancement.

### To do

(Include the to-do list for this PR to be finished here.)

* [ ] Tests
* [ ] Documentation
* [ ] Feature flags
* [ ] …

### Related issues

Would address #1300.

iay commented:
1. I might take a look this weekend at adding some tests for this, it's about time I learned TDD in Ruby. What's the best reference for the testing framework you're using?

2. I'm not sure how to interpret the CI failures, are they on my side somehow?

3. Do you want me to rebase this against the current HEAD?

iay commented:
I think I rebased the original commit; CI still failing.

ddfreyne commented:
@iay The only remaining error that I can see is

```
 1) Failure:
Nanoc::Helpers::BloggingTest#test_atom_feed_without_base_url [/home/travis/build/nanoc/nanoc/nanoc/test/helpers/test_blogging.rb:215]:
[Nanoc::Int::Errors::GenericTrivial] exception expected, not
Class: <NoMethodError>
Message: <"undefined method `+' for nil:NilClass">
---Backtrace---
/home/travis/build/nanoc/nanoc/nanoc/lib/nanoc/helpers/blogging.rb:196:in `atom_feed'
/home/travis/build/nanoc/nanoc/nanoc/test/helpers/test_blogging.rb:216:in `block in test_atom_feed_without_base_url'
---------------
```

… which I believe is related to the change made

iay commented:
Thanks; I need to fix that but I'll do it in conjunction with learning RSpec and adding some positive tests for the new code. Should get to it over the next couple of weeks (there's really not hurry on this one).

iay commented:
Reworked to avoid the exception, rebased on latest.

I'm trying to figure out how I would make tests for this change. Am I right in saying that you have both RSpec based tests under `nanoc/spec` and also more conventional unit tests under `nanoc/test`?

It looks like there are no tests for the actual XML build in the RSpec section of the repository, so I'm guessing that it might be easiest to add some tests to the end of `nanoc/test/helpers/test_blogging.rb` and just match expected content within the XML for the item as a string, as for example is being done in for example `test_atom_feed_with_xml_base`. Does that sound right?

iay commented:
A related question would be whether there was an XML parser available in that test context; it seems like a test that re-parsed the resulting string and then dug into the structure would be less brittle than one involving just regexes.

ddfreyne commented:
> Am I right in saying that you have both RSpec based tests under `nanoc/spec` and also more conventional unit tests under `nanoc/test`?

Correct. The `test/` directory is older; eventually I hope to have everything migrated to `spec/`, but there’s no rush.

> I'm guessing that it might be easiest to add some tests to the end of `nanoc/test/helpers/test_blogging.rb` and just match expected content within the XML for the item as a string, as for example is being done in `test_atom_feed_with_xml_base`. Does that sound right?

Yes!

ddfreyne commented:
> A related question would be whether there was an XML parser available in that test context; it seems like a test that re-parsed the resulting string and then dug into the structure would be less brittle than one involving just regexes.

Definitely. Nokogiri likely will be available, so you could do something similar to what’s being done in `test_xml_sitemap.rb`:

```ruby
      doc = Nokogiri::XML(res)
      urlsets = doc.css('> urlset')
      assert_equal 1, urlsets.size
      urls = urlsets.css('> url')
      assert_equal 4, urls.size
      assert_equal 'http://example.com/item-one/a/',   urls[0].css('> loc').inner_text
      assert_equal 'http://example.com/item-one/b/',   urls[1].css('> loc').inner_text
      assert_equal 'http://example.com/item-three/a/', urls[2].css('> loc').inner_text
      assert_equal 'http://example.com/item-three/b/', urls[3].css('> loc').inner_text
      # …
```

iay commented:
Added unit tests.

So now I know how to use Nokogiri, I guess. Is there anything else you need here?

Also, is there a way to manually run just a single file of tests in isolation? Took a while to get started on this because it takes a while to run everything just to find out that I've written my XPath expressions incorrectly 😄

ddfreyne commented:
> Also, is there a way to manually run just a single file of tests in isolation?

Yeah, but I just realised that this approach isn’t documented.

From within the `nanoc` subdirectory, either

* (for something in `test`) run `bundle exec m test/path/to/test_something.rb`
* (for something in `spec`) run `bundle exec rspec spec/path/to/something_spec.rb`
